server/src/main/java/org/opensearch/index/query/CombinedFieldQueryBuilder.java

```diff
@@ -0,0 +1,335 @@
+        return zeroTermsQuery;
+    }
+
+    static boolean isTextField(MappedFieldType fieldType) {
+        if (fieldType == null) {
+            return false;
+        }
+        return fieldType.getTextSearchInfo() != TextSearchInfo.NONE && fieldType.getTextSearchInfo() != TextSearchInfo.SIMPLE_MATCH_ONLY;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        boolean hasTextField = fieldWeights.keySet().stream().anyMatch(k -> isTextField(context.fieldMapper(k)));
+        if (hasTextField == false) {
+            return Queries.newUnmappedFieldsQuery(fieldWeights.keySet());
+        }
+        Analyzer explicitAnalyzer = null;
+        if (analyzer != null) {
+            explicitAnalyzer = context.getMapperService().getIndexAnalyzers().get(analyzer);
+            if (explicitAnalyzer == null) {
+                throw new IllegalArgumentException("No analyzer found for [" + analyzer + "]");
+            }
+        }
+
+        CombinedFieldQuery.Builder builder = new CombinedFieldQuery.Builder();
+        boolean hasTerms = false;
+        for (Map.Entry<String, Float> fieldWeight : fieldWeights.entrySet()) {
+            String fieldName = fieldWeight.getKey();
+            MappedFieldType fieldType = context.fieldMapper(fieldName);
+            if (isTextField(fieldType) == false) {
+                // ignore unmapped fields or fields that do not support text search
+                continue;
+            }
+            builder.addField(fieldName, fieldWeight.getValue());
+
+            Analyzer fieldAnalyzer;
+            if (explicitAnalyzer == null) {
+                // Use per-field analyzer
+                fieldAnalyzer = context.getSearchAnalyzer(fieldType);
+            } else {
+                fieldAnalyzer = explicitAnalyzer;
+            }
+            hasTerms = collectAllTerms(fieldName, fieldAnalyzer, value.toString(), builder) || hasTerms;
+        }
+        if (hasTerms == false) {
+            switch (zeroTermsQuery) {
+                case NONE:
+                    return Queries.newMatchNoDocsQuery("Matching no documents because no terms present");
+                case ALL:
+                    return Queries.newMatchAllQuery();
+                default:
+                    throw new IllegalStateException("unknown zeroTermsQuery " + zeroTermsQuery);
+            }
+        }
+        return builder.build();
+    }
+
+    private static boolean collectAllTerms(String fieldName, Analyzer analyzer, String queryString, CombinedFieldQuery.Builder builder)
+        throws IOException {
+        boolean hasTerms = false;
+        TokenStream tokenStream = analyzer.tokenStream(fieldName, queryString);
+        TermToBytesRefAttribute termAtt = tokenStream.addAttribute(TermToBytesRefAttribute.class);
+        tokenStream.reset();
+        while (tokenStream.incrementToken()) {
+            builder.addTerm(BytesRef.deepCopyOf(termAtt.getBytesRef()));
```

Comment:
This is super interesting (not sure if it is good or bad), since each field may use own analyzer, we may end up in a situation when index time analyzer != search time analyzer: the field may have a match for a term produced by other field analyzer (probably stemming is a good example). Do you think it is an issue here?

Reply:
That is definitely a possibility. 

I can imagine some convoluted scenarios with custom analyzers that lead to "clearly" incorrect results. Say you have an analyzer that outputs a single "true" or "false" token if the input (index-time or query-time) contains the word "foo". If your query is "here is a true statement" and you combine that has_foo field, you would match everything where `has_foo:true` (even though the query doesn't have "foo"). 

It's a really bad example, but I don't have a good example off the top of my head.

On the flip side, I'm thinking of the following arguments:
1. The purely disjunctive nature of CombinedFieldQuery feels to me like it's going for "Match 'em all and let scoring sort 'em out".
2. If someone does need to restrict to "valid" hits, they can add the equivalent `multi_match` query as a filter.

Oh -- actually, if we apply that second point to my bad example from above, you can imagine a scenario where you're *not* matching on `has_foo:true` (thanks to the well-behaved `multi_match` filter), but you are increasing the score of a match with `has_foo:true` (because of the badly-behaved `combined_field` query), even though the query doesn't have "foo". So, the scoring would be wrong, which basically breaks my first argument.

Based on that thought process, I'm going to change it to require all fields to use the same search analyzer.

At that point, does it really make sense to support a custom `analyzer` parameter? Since the query needs to contain it explicitly, I feel like I'm okay with people explicitly shooting themselves in the foot.

Reply:
> Based on that thought process, I'm going to change it to require all fields to use the same search analyzer.

Very much agree on that, thank you

> At that point, does it really make sense to support a custom analyzer parameter? Since the query needs to contain it explicitly, I feel like I'm okay with people explicitly shooting themselves in the foot.

I think this is fine and it make sense to support that, as far as it is applied to all fields - the scope of what could go wrong is significantly smaller, thank you.

Reply:
Oh, continuing my contrived example from above, imagine the doc frequency of "true" in the text field is high, but the doc frequency of "false" is low. 

A search for "true" would produce a term "false" (from the `has_foo` analyzer), match the term "false" on the text field, and probably score it highly because it's a rare term. So a "false" document would probably be the top match for a "true" search.

It's a silly example, but it's convinced me that my previous approach was totally wrong. 😄